### PR TITLE
adding to repository on behalf of another programmer

### PR DESCRIPTION
**Pull request by paulsj-80**

### CODE REVIEW

File: modules/utils/graph/Redif.hpp [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+#include "3rdparty/blaze/Blaze.h"
+#include "modules/distance.hpp"
+
+namespace metric
+{
+	template<typename recType, class Metric = metric::Euclidian<recType>, typename Tv = double>
+	class Redif
+	{
+	public:
+		Redif(std::vector<std::vector<Tv>> trainData, size_t nNeighbors = 10, size_t nIter = 15);
```
Comment (mkh-welsch) [inline modules/utils/graph/Redif.hpp]: nNeibhbors = kNeighbors = 7

File: modules/utils/graph/Redif.hpp [hunk 2]
```diff
@@ -0,0 +1,35 @@
+#pragma once
+
+#include <vector>
+
+#include "3rdparty/blaze/Blaze.h"
```
Comment (mkh-welsch) [inline modules/utils/graph/Redif.hpp]: update to correct relativ path

File: modules/utils/graph/Redif.cpp [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+	size_t Redif<recType, Metric, Tv>::partition(blaze::DynamicVector<size_t> & idx, blaze::DynamicVector<Tv, flag> & val, size_t start, size_t end)
+	{
+		double pivot = val[end];
+
+		size_t P_index = start;
+		size_t i;
+		size_t idxTemp;
+		double valTemp;
+
+		for (i = start; i < end; i++)
```
Comment (mkh-welsch) [inline modules/utils/graph/Redif.cpp]: do not define index variables outside the loop, when not really necessary

File: modules/utils/graph/Redif.cpp [hunk 4]
```diff
@@ -0,0 +1,325 @@
...
+			blaze::DynamicMatrix<Tv> graphLap = calcWeightedGraphLaplacian(localDistMat);
+			LArray.push_back(graphLap);
+			blaze::DynamicMatrix<Tv> D(n, n);
+			for (size_t i = 0; i < D.rows(); i++)
+				D(i, i) = graphLap(i, i);
+			blaze::DynamicMatrix<Tv> invLeft(n, n);
+			invLeft = blaze::inv(D + 0.25*graphLap);
+			xTrainDe = invLeft * (D*xTrainDe);
+
+			//cout << endl << endl << endl << endl << endl << endl;
```
Comment (mkh-welsch) [inline modules/utils/graph/Redif.cpp]: remove debug stuff

File: modules/utils/graph/Redif.cpp [hunk 5]
```diff
@@ -0,0 +1,325 @@
...
+				ixx(i, j) = j;
+
+		blaze::DynamicMatrix<Tv> dist(n, n);
+		for (size_t i = 0; i < n; i++)
+			for (size_t j = 0; j < n; j++)
+			{
+				if (i == j)
+				{
+					dist(i, j) = 0;
+					continue;
```
Comment (mkh-welsch) [inline modules/utils/graph/Redif.cpp]: make clear if else structures without jumps

File: modules/utils/graph/Redif.cpp [hunk 6]
```diff
@@ -0,0 +1,325 @@
...
+		size_t n = dataSample.rows();
+
+		blaze::DynamicMatrix<size_t> ixx(nNeighbors, n);
+		for (size_t i = 0; i < nNeighbors; i++)
+			for (size_t j = 0; j < n; j++)
+				ixx(i, j) = j;
+
+		blaze::DynamicMatrix<Tv> dist(n, n);
+		for (size_t i = 0; i < n; i++)
+			for (size_t j = 0; j < n; j++)
```
Comment (mkh-welsch) [inline modules/utils/graph/Redif.cpp]: distance matrix is symmetric. 100% overhead here in computation, use distance matrix class in space module instead.

File: modules/utils/graph/Redif.cpp [hunk 7]
```diff
@@ -0,0 +1,325 @@
...
+		for (size_t i = 0; i < n; i++)
+		{
+			blaze::DynamicVector<size_t> tempIdx(n);
+			blaze::DynamicVector<Tv> tempDist(n);
+			for (size_t j = 0; j < n; j++)
+			{
+				tempIdx[j] = j;
+				tempDist[j] = dist(j, i);
+			}
+			std::cout << i << "-th" << std::endl;
```
Comment (mkh-welsch) [inline modules/utils/graph/Redif.cpp]: remove debug info or integrate "verbose" flag during development

File: modules/utils/graph/Redif.cpp [hunk 8]
```diff
@@ -0,0 +1,325 @@
...
+			Quicksort(tempIdx, tempDist, 0, n - 1);
+			//if (i == 58)
+			//{
+			//	std::cout << std::endl;
+			//	for (size_t j = 0; j < 30; j++)
+			//		std::cout << tempDist[j] << std::endl;
+			//}
```
Comment (mkh-welsch) [inline modules/utils/graph/Redif.cpp]: no debug

File: modules/utils/graph/Redif.cpp [hunk 9]
```diff
@@ -0,0 +1,325 @@
...
+
+		/// Degree matrix
+		blaze::DynamicMatrix<Tv> D(n, n);
+		for (size_t i = 0; i < D.rows(); i++)
+		{
+			double sum = 0;
+			for (size_t j = 0; j < n; j++)
+				sum += localDist(i, j);
+			if (sum == 0) sum = 1;
+			D(i, i) = 1 / sum;
```
Comment (mkh-welsch) [inline modules/utils/graph/Redif.cpp]: beware of plain numbers without types in combination with templates. They are always integers. make Tv(1) for example. and check, if it really works with integers. implements some checks for incompatible types if_integral .... etc.

File: modules/utils/graph/Redif.cpp [hunk 10]
```diff
@@ -0,0 +1,325 @@
...
+	}
+
+	template <typename recType, class Metric, typename Tv>
+	blaze::DynamicMatrix<Tv> Redif<recType, Metric, Tv>::encode(blaze::DynamicMatrix<Tv> x, blaze::DynamicVector<Tv> & l_idx)
+	{
+		size_t nTrain = xTrain.rows();
+		size_t nX = x.rows();
+		size_t nIter = LArray.size();
+		/// Find nearest Neighbors of each record of x in xTrain
+		size_t l = 1; /// Number of nearest neighbors
```
Comment (mkh-welsch) [inline modules/utils/graph/Redif.cpp]: l ?

File: modules/utils/graph/Redif.hpp [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+{
+	template<typename recType, class Metric = metric::Euclidian<recType>, typename Tv = double>
+	class Redif
+	{
+	public:
+		Redif(std::vector<std::vector<Tv>> trainData, size_t nNeighbors = 10, size_t nIter = 15);
+		~Redif() = default;
+		
+		blaze::DynamicMatrix<Tv> encode(blaze::DynamicMatrix<Tv> x, blaze::DynamicVector<Tv> & l_idx);
+		blaze::DynamicMatrix<Tv> decode(blaze::DynamicMatrix<Tv> xEncoded, blaze::DynamicMatrix<Tv> xTrainEncoded, blaze::DynamicVector<size_t> l_idx);
```
Comment (mkh-welsch) [inline modules/utils/graph/Redif.hpp]: very very wrong interface: do not use other inputs then x for encode and xEncode for decode methods. the internals must be hided in public methods.

File: modules/utils/graph/Redif.hpp (hunk 1 above)
Comment (mkh-welsch) [inline modules/utils/graph/Redif.hpp]: wrong interface. separate the type factory function with metric and structural parameters as input from pseudo constructor with data set as input

File: modules/utils/graph/Redif.cpp [hunk 12]
```diff
@@ -0,0 +1,325 @@
...
+		idxTemp = idx[end];
+		idx[end] = idx[P_index];
+		idx[P_index] = idxTemp;
+
+		return P_index;
+	}
+
+	template <typename recType, class Metric, typename Tv>
+	template<bool flag>
+	void Redif<recType, Metric, Tv>::Quicksort(blaze::DynamicVector<size_t> & idx, blaze::DynamicVector<Tv, flag> & data, size_t start, size_t end)
```
Comment (mkh-welsch) [inline modules/utils/graph/Redif.cpp]: why own quicksort? it is incredible slow, when implementing on its own in comparison to STL version, and it blows the code size.

File: modules/utils/graph/Redif.cpp [hunk 13]
```diff
@@ -0,0 +1,325 @@
...
+		for (size_t i = 0; i < nX; i++)
+			for (size_t j = 0; j < xEncoded.columns(); j++)
+				xEncoded(l_idx[i], j) = x(i, j);
+
+		for (size_t i = 0; i < nIter; i++)
+		{
+			blaze::DynamicMatrix<Tv> D(nTrain, nTrain);
+			for (size_t k = 0; k < D.rows(); k++)
+				D(k, k) = LArray[i](k, k);
+			xEncoded = blaze::inv(D + 0.25*LArray[i])*(D*xEncoded);
```
Comment (mkh-welsch) [inline modules/utils/graph/Redif.cpp]: no magic numbers (0.25) when it is no magic number, leave comment here

File: modules/utils/graph/Redif.hpp [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+
+		blaze::DynamicMatrix<Tv> getLocalDistMatrix(blaze::DynamicMatrix<Tv> dataSample);
+		blaze::DynamicMatrix<Tv> calcWeightedGraphLaplacian(blaze::DynamicMatrix<Tv> localDist);
+		void trainModel(blaze::DynamicMatrix<Tv> & xTrainDe, size_t nIter);
+
+		blaze::DynamicMatrix<Tv> xTrain;
+		std::vector<blaze::DynamicMatrix<Tv>> LArray;
+		size_t nNeighbors;
+	};
+}
```
Comment (mkh-welsch) [inline modules/utils/graph/Redif.hpp]: example file is missing.

Comment (mkh-welsch) [inline modules/utils/graph/Redif.hpp]: #include "Rediff.cpp" is missing here to work header only

File: modules/utils/graph/Redif.hpp [hunk 15]
```diff
@@ -0,0 +1,35 @@
+#pragma once
```
Comment (mkh-welsch) [inline modules/utils/graph/Redif.hpp]: make classical header guards, so one can check includes later better

File: modules/utils/graph/Redif.cpp [hunk 16]
```diff
@@ -0,0 +1,325 @@
...
+			//cout << endl << endl << endl << endl << endl << endl;
+			//cout << i << "-th" << endl;
+			//for (size_t j = 0; j < 10; j++)
+			//	cout << graphLap(0, j) << "\t";
+		}
+
+	}
+
+	template <typename recType, class Metric, typename Tv>
+	blaze::DynamicMatrix<Tv> Redif<recType, Metric, Tv>::getLocalDistMatrix(blaze::DynamicMatrix<Tv> dataSample)
```
Comment (mkh-welsch) [inline modules/utils/graph/Redif.cpp]: why no constant reference?

File: modules/utils/graph/Redif.hpp (hunk 14 above)
Comment (mkh-welsch) [reply modules/utils/graph/Redif.hpp]: benchmark is missing

File: modules/utils/graph/Redif.cpp [hunk 17]
```diff
@@ -0,0 +1,325 @@
+#include "Redif.hpp"
+
+
+namespace metric
+{
+	template <typename recType, class Metric, typename Tv>
+	Redif<recType, Metric, Tv>::Redif(std::vector<std::vector<Tv>> trainData, size_t nNeighbors /* = 10 */, size_t nIter /* = 15 */)
+	{
+		this->nNeighbors = nNeighbors;
```
Comment (mkh-welsch) [inline modules/utils/graph/Redif.cpp]: use initializer list instead

**Comment by paulsj-80** [comment]
Closing pull request as it doesn't fit. Would suggest to not delete branch itself, as it may be of use.

**Pull request closed without merging**